app/client/components/duplicateWidget.ts

```diff
@@ -0,0 +1,318 @@
+import {IPageWidget, toPageWidget} from 'app/client/ui/PageWidgetPicker';
+import {select} from 'app/client/ui2018/menus';
+import {saveModal} from 'app/client/ui2018/modals';
+import {BulkColValues, getColValues, RowRecord, UserAction} from 'app/common/DocActions';
+import {arrayRepeat} from 'app/common/gutil';
+import {schema} from 'app/common/schema';
+import {dom, Observable} from 'grainjs';
+import cloneDeepWith from 'lodash/cloneDeepWith';
+import flatten from 'lodash/flatten';
+import forEach from 'lodash/forEach';
+import {testId} from 'app/client/ui2018/cssVars';
+import {logTelemetryEvent} from 'app/client/lib/telemetry';
+import sortBy from 'lodash/sortBy';
+import {fromPairs} from 'lodash';
+
+const t = makeT('duplicateWidget');
+
+// Duplicate page with pageId. Starts by prompting user for a new name.
+export async function buildDuplicateWidgetModal(gristDoc: GristDoc, viewSectionId: number) {
+  const activeView = gristDoc.activeViewId.get();
+  const pages = sortBy(gristDoc.docModel.pages.rowModels.filter(p => p), [(row) => row.pagePos.peek()]);
+  const pageSelectOptions = pages.map(page => {
+    const view = page.view.peek();
+    const isActivePage = view.getRowId() === activeView;
+    const suffix = isActivePage? " (Active)" : "";
+    return {
+      label: `${view.name.peek()}${suffix}`,
+      value: view.getRowId(),
+      isActivePage: isActivePage,
+    };
+  });
+
+  pageSelectOptions.push({ label: 'Create new page', value: 0, isActivePage: false });
+
+  // Logically this should never happen, as a Grist doc without pages should be impossible.
+  if (pageSelectOptions.length < 1) {
+    throw new Error("No pages available to duplicate widget to");
+  }
+
+  saveModal((ctl, owner) => {
+    const initialSelectedPage =
+      pageSelectOptions.find(option => option.isActivePage)?.value ?? pageSelectOptions[0].value;
+    const pageSelectObs = Observable.create<number>(owner, initialSelectedPage);
+
+    return {
+      title: t('Duplicate widget'),
+      body: dom('div', [
+        cssField(
+          cssLabel("Page"),
+          select(pageSelectObs, pageSelectOptions),
+          testId("duplicate-widget-page-select"),
+        )
+      ]),
+      async saveFunc() {
+        await duplicateWidgets(
+          gristDoc, [viewSectionId], pageSelectObs.get()
+        );
+      }
+    };
+  });
+}
+
+export async function duplicateWidgets(gristDoc: GristDoc, srcViewSectionIds: number[], destViewId: number) {
+  const allViewSectionModels = gristDoc.docModel.viewSections.rowModels;
+  const validWidgetIds = srcViewSectionIds.filter(id => allViewSectionModels[id]);
+
+  // Generally this shouldn't happen, but it catches a theoretically possible edge case.
+  if (validWidgetIds.length === 0) {
+    throw new Error("Unable to duplicate widgets as no valid source widget IDs were provided");
+  }
+
+  const sourceViewSections = validWidgetIds.map(id => allViewSectionModels[id]);
```

Comment:
Maybe you can merge this with the line in 71, to run this filter only once (and the `validWidgetIds` is not needed afterward)

Reply:
Done - double check you're happy with it? :)